lib/maths/time_series/CCalendarCyclicTest.cc

```diff
@@ -313,33 +400,49 @@ double CCalendarCyclicTest::survivalFunction(double error) const {
     try {
         boost::math::normal normal{common::CBasicStatistics::mean(tailMoments),
                                    std::sqrt(common::CBasicStatistics::variance(tailMoments))};
-        return (100.0 - LARGE_ERROR_PERCENTILE) / 100.0 *
-               common::CTools::safeCdfComplement(normal, error);
+        return (100.0 - this->largeErrorPercentile()) / 100.0 *
+               common::CTools::safeCdfComplement(normal, this->errorAdj(error));
 
     } catch (const std::exception& e) {
         LOG_ERROR(<< "Failed to compute tail distribution '" << e.what() << "'");
     }
     return 1.0;
 }
 
-double CCalendarCyclicTest::significance(double n, double nl, double nv) const {
-    if (n > 0.0) {
-        try {
-            // We have roughly 31 independent error samples, one for each
-            // day of the month, so the chance of seeing as extreme an event
-            // among all of them is:
-            //   1 - P("don't see as extreme event") = 1 - (1 - P("event"))^31
-            boost::math::binomial bl{n, 1.0 - LARGE_ERROR_PERCENTILE / 100.0};
-            boost::math::binomial bv{n, 1.0 - VERY_LARGE_ERROR_PERCENTILE / 100.0};
-            double p{std::min({2.0 * common::CTools::safeCdfComplement(bl, nl - 1.0),
-                               2.0 * common::CTools::safeCdfComplement(bv, nv - 1.0), 1.0})};
-            return common::CTools::oneMinusPowOneMinusX(p, 31.0);
-        } catch (const std::exception& e) {
-            LOG_ERROR(<< "Failed to calculate significance: " << e.what()
-                      << " n = " << n << " nl = " << nl << " nv = " << nv);
-        }
-    }
-    return 1.0;
+double CCalendarCyclicTest::errorsPValue(double n, double nl, double nv) const {
+    return n > 0.0
+               ? std::min(
+                     binomialPValueAdj(n, 1.0 - this->largeErrorPercentile() / 100.0, nl),
+                     binomialPValueAdj(n, 1.0 - this->veryLargeErrorPercentile() / 100.0, nv))
+               : 1.0;
+}
+
+double CCalendarCyclicTest::sufficientCountToMeasureLargeErrors() const {
+    // Cap the how long we'll wait identify large errors.
+    return std::min(static_cast<double>(20 * core::constants::DAY) / m_BucketLength, 100.0);
+}
+
+double CCalendarCyclicTest::largeErrorPercentile() const {
+    // To test for calendar components we check if they align with excess
+    // counts in the right tail of the error distribution. For long bucket
+    // lengths we expect to see few high percentile errors in the window
+    // as a whole and the expected variance of high percentile counts is
+    // too large to be useful. Instead we adjust the threshold above which
+    // we count errors so the expected counts are fixed for long bucket
+    // lengths.
+    return 100.0 - (100.0 - LARGE_ERROR_PERCENTILE) *
```

Comment:
Does it make sense to factor out the code that repeats in `largeErrorPercentile` and `veryLargeErrorPercentile`?